// @remotion/studio-shared/dist/optimistic-add-keyframe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optimisticAddEffectKeyframe = exports.optimisticAddSequenceKeyframe = void 0;
const keyframe_interpolation_function_1 = require("./keyframe-interpolation-function");
const addKeyframeToPropStatus = ({ status, fieldKey, frame, value, schema, }) => {
    var _a;
    if (status.status === 'keyframed') {
        const existingIndex = status.keyframes.findIndex((kf) => kf.frame === frame);
        if (existingIndex !== -1) {
            const updatedKeyframes = status.keyframes.map((keyframe, index) => index === existingIndex ? { frame, value } : keyframe);
            return {
                ...status,
                keyframes: updatedKeyframes,
            };
        }
        const keyframes = [...status.keyframes, { frame, value }].sort((first, second) => first.frame - second.frame);
        const easing = [...status.easing];
        while (easing.length < keyframes.length - 1) {
            easing.push('linear');
        }
        return {
            ...status,
            keyframes,
            easing,
        };
    }
    if (status.status === 'static') {
        const staticValue = (_a = status.codeValue) !== null && _a !== void 0 ? _a : value;
        return {
            status: 'keyframed',
            interpolationFunction: (0, keyframe_interpolation_function_1.getKeyframeInterpolationFunction)({
                schema,
                key: fieldKey,
                staticValue,
                newValue: value,
            }),
            keyframes: [{ frame, value }],
            easing: [],
            clamping: { left: 'clamp', right: 'clamp' },
            posterize: undefined,
        };
    }
    return status;
};
const optimisticAddSequenceKeyframe = ({ previous, fieldKey, frame, value, schema, }) => {
    if (!previous.canUpdate) {
        return previous;
    }
    if (!(0, keyframe_interpolation_function_1.isSchemaFieldKeyframable)({ schema: schema !== null && schema !== void 0 ? schema : null, key: fieldKey })) {
        return previous;
    }
    const status = previous.props[fieldKey];
    if (!status) {
        return previous;
    }
    return {
        ...previous,
        props: {
            ...previous.props,
            [fieldKey]: addKeyframeToPropStatus({
                status,
                fieldKey,
                frame,
                value,
                schema: schema !== null && schema !== void 0 ? schema : null,
            }),
        },
    };
};
exports.optimisticAddSequenceKeyframe = optimisticAddSequenceKeyframe;
const optimisticAddEffectKeyframe = ({ previous, effectIndex, fieldKey, frame, value, schema, }) => {
    if (!previous.canUpdate) {
        return previous;
    }
    if (!(0, keyframe_interpolation_function_1.isSchemaFieldKeyframable)({ schema: schema !== null && schema !== void 0 ? schema : null, key: fieldKey })) {
        return previous;
    }
    const targetIndex = previous.effects.findIndex((e) => e.effectIndex === effectIndex);
    if (targetIndex === -1) {
        return previous;
    }
    const target = previous.effects[targetIndex];
    if (!target.canUpdate) {
        return previous;
    }
    const status = target.props[fieldKey];
    if (!status) {
        return previous;
    }
    const updatedEffect = {
        ...target,
        props: {
            ...target.props,
            [fieldKey]: addKeyframeToPropStatus({
                status,
                fieldKey,
                frame,
                value,
                schema: schema !== null && schema !== void 0 ? schema : null,
            }),
        },
    };
    const effects = [...previous.effects];
    effects[targetIndex] = updatedEffect;
    return {
        ...previous,
        effects,
    };
};
exports.optimisticAddEffectKeyframe = optimisticAddEffectKeyframe;
